tools/generate_config.py

```diff
@@ -0,0 +1,326 @@
+                var_name = 'key_len'
+              header.write('    size_t %s;\n' % var_name)
+            header.write('    %s %s%s;\n' % (pre_type(jtype), name, post_type(jtype)))
+            num_settings += 1
+          num_settings -= 1
+          header.write('} %s;\n\n' % typename)
+
+    header.write('extern const sgxlkl_enclave_config_t sgxlkl_default_enclave_config;\n\n');
+
+    header.write(
+      "typedef struct {\n"
+      "    char* scope;\n"
+      "    char* type;\n"
+      "    char* description;\n"
+      "    char* default_value;\n"
+      "    char* override_var;\n"
+      "} sgxlkl_enclave_setting_t;\n\n");
+
+    header.write("extern const sgxlkl_enclave_setting_t sgxlkl_enclave_settings[%d];\n\n" % num_settings);
+
+    header.write('#endif /* SGXLKL_ENCLAVE_CONFIG_H */');
+
+source_includes = """
+#ifdef SGXLKL_ENCLAVE
+#include <enclave/enclave_util.h>
+#include <enclave/oe_compat.h>
+#define FAIL sgxlkl_fail
+#else
+#include <host/sgxlkl_util.h>
+#include <string.h>
+#define FAIL sgxlkl_host_fail
+#endif
+"""
+
+def generate_source(schema_file_name, root, args):
+  with open(str(args.source), "w") as source:
+    source.write('/* Automatically generated from %s; do not modify. */\n' % schema_file_name);
+
+    source.write(source_includes)
+    source.write('\n')
+
+    source.write('#include "%s"\n\n' % args.header);
+
+    # enum conversions
+    for typename, typedef in root['definitions'].items():
+      if 'enum' in typedef:
+        if 'c_enum' in typedef:
+          names = typedef['enum']
+          c_names = typedef['c_enum']
+          if len(names) != len(c_names):
+            raise Exception("ERROR: length of c_enum does not match enum in %s" % typename)
+          source.write('const char* %s_to_string(%s e)\n{\n' % (typename, typename))
+          source.write('  switch(e) {\n')
+          for i in range(len(names)):
+            name = names[i]
+            c_name = c_names[i]
+            source.write('    case %s: return "%s";\n' % (c_name, name))
+          source.write('    default: return ""; /* Unreachable */\n')
+          source.write('  }\n')
+          source.write('}\n\n')
+          source.write('%s string_to_%s(const char *e)\n{\n' % (typename, typename))
+          for i in range(len(names)):
+            name = names[i]
+            c_name = c_names[i]
+            source.write('  if (strcmp(e, "%s") == 0) return %s;\n' % (name, c_name))
+          source.write('  FAIL("unknown enum value \'%s\'\\n", e);\n')
+          source.write('  return %s;\n\n' % (c_names[0]))
+          source.write('}\n\n')
+
+
+    # default config
+    source.write('const sgxlkl_enclave_config_t sgxlkl_default_enclave_config = {\n')
+    scope = []
+    def initialize(scope, elem):
+      indent = '    ' * (len(scope) + 1)
+      typedef = root['definitions'][elem]
+      if 'enum' not in typedef:
+        for name, jtype in typedef['properties'].items():
+          if name == 'format_version':
```

Comment:
Why is this handled separately? Couldn't you just check if no default exists?

Reply:
It's not about the default. This just tracks the version number of the schema/structs. The host adds this to the json configs and the enclave reads it only to check it supports it. It's never saved as an actual config setting in any of the structs, i.e. it should be in the schema (if we want to validate them), but not `sgxlkl_enclave_config_t` or in the default config.

Reply:
Just occurred to me: I'm sure JSON schemas have a concept of versioning? If so, we should really use their way of checking versions.

Reply:
They don't, but I still don't understand. The host and enclave always come in a pair, so there's no point in assuming that the host comes from version x and the enclave from version y. The version has to be put in by the user, not the launcher. And then it becomes a normal field with a default of whatever the current version is, maybe even without a default and making it required. If this is too much for this PR then we can also remove the version field again and do it later.

Reply:
I'm fine with that (in fact happy to kill those code paths). I think your original argument was that this allows us to upgrade the infrastructure (i.e. new launcher) but we won't have to re-package other's containers (i.e. old image). 

The thought of a user-editable version field never crossed my mind. Do you think users would want that sort of setting to keep track of their container versions? They can easily compile that into their application code, so I don't think we should add that to the config.

Reply:
Now I finally understand what you mean by image: the Docker image containing the disk image files and configuration files. The configuration version in SGX-LKL config files is not the same as versions of applications or Docker images and hence it doesn't make sense to embed that into application code. The point of including a schema version in the user-editable config files is to give the launcher a chance to migrate old configuration schemas to new ones or simply reject it if there was a breaking change. The point is that it's not a good idea to just hope for the best.

Reply:
Ah I see the confusion now. I was talking about the enclave image (that's what I mainly care about) and I assumed you were talking about the same kind of image when you were really talking about container images which include both, the launcher and the enclave image, so we can assume they're always of the same version. I will kill the corresponding branches in the enclave code, I never liked that either, but we can keep the version number with a default.

Reply:
Could you please read what I've written again? The Docker image does **not** contain the launcher or the enclave image.

Reply:
It makes no difference whatsover. 

(Quote: "the Docker image containing...", Quote: "The Docker image does not contain...". It's fine if Docker != Container, they are the same to me.)